app/models/vm/operations.rb

```diff
@@ -15,14 +17,16 @@ module Vm::Operations
 
   def cockpit_url
     return if ipaddresses.blank?
-    miq_server = ext_management_system.nil? ? nil : ext_management_system.zone.remote_cockpit_ws_miq_server
-    MiqCockpit::WS.url(miq_server,
-                       MiqCockpitWsWorker.fetch_worker_settings_from_server(miq_server),
-                       ipv4_address || ipaddresses.first)
+    MiqCockpit::WS.url(cockpit_server, cockpit_worker, ipv4_address || ipaddresses.first)
   end
 
   def ipv4_address
-    ipaddresses.find { |ip| IPAddr.new(ip).ipv4? }
+    return public_address unless public_address.nil?
+    ipaddresses.find { |ip| IPAddr.new(ip).ipv4? unless ip.starts_with?('192') }
```

Comment:
Is this robust enough? I think i've seen 10.x IPs on amazon.

Does the VM have any notion if it's hostname is public or not? If it did that might make this simpler.

Reply:
@petervo  Line 25 allows for 10.x ip addresses in my testing (192.x addresses have been unreachable).  This solution will even try 192.x ip addresses (unreachable in my testing) in the case the `ipv4_address` method returns nil in the `cockpit_url` method.

I'm trying for the lowest common denominator and have verified that we can connect to Amazon, GCE, RHOS, RHEV and VMware virtual machines successfully.

If this isn't robust enough for more complex Amazon or <insert-provider-here>, it can be addressed via provider overrides in a follow-up PR per @Fryguy's comment above [1]. 

[1]  https://github.com/ManageIQ/manageiq/pull/15901#pullrequestreview-59538858

Reply:
I meant 10.x private IPs which in theory would not be reachable.

But I agree this covers most use-cases and like you said it can always be adjusted later if needed

Reply:
So, this is trying 10.0.0.0/8, 172.16.0.0/12, 192.168.0.0/16 if they're the only ipaddresses on the system?

Can this be: 
```ruby
ipaddresses.find { |ip| IPAddr.new(ip).ipv4? && !ip.starts_with?('192') }
```

Reply:
@jrafanie, Yes, the `ipv4_address` tries for the most common scenarios (public to private) and falls back to `ipaddresses.first` if nothing matches.  All 3 ip addresses above would be tried on their own (experienced while debugging and testing this PR...).

Updated the code per your request above, thanks!

Reply:
Other than going to some remote system to ask what your public ip address is, I'm not sure what else we can do.  I'm good with the changes as it seems to cover most of the bases, until it doesn't and we'll worry about it then.